core/bloom.go

```diff
@@ -0,0 +1,358 @@
+// with default parameters.
+func evalBFAdd(args []string) []byte {
+	if len(args) != 2 {
+		return Encode(fmt.Errorf("%w for 'BFADD' command", errWrongArgs), false)
+	}
+
+	var key string = args[0]
+	opts, _ := newBloomOpts(args[1:], true)
+
+	bloom, err := getOrCreateBloomFilter(key, opts)
+	if err != nil {
+		return Encode(fmt.Errorf("%w for 'BFADD' command", err), false)
+	}
+
+	resp, err := bloom.add(args[1])
+	if err != nil {
+		return Encode(fmt.Errorf("%w for 'BFADD' command", err), false)
+	}
+
+	return resp
+}
+
+// evalBFExists evaluates the BFEXISTS command responsible for checking existance
+// of an element in a bloom filter.
+func evalBFExists(args []string) []byte {
+	if len(args) != 2 {
+		return Encode(fmt.Errorf("%w for 'BFEXISTS' command", errWrongArgs), false)
+	}
+
+	var key string = args[0]
+	bloom, err := getOrCreateBloomFilter(key, nil)
+	if err != nil {
+		return Encode(fmt.Errorf("%w for 'BFEXISTS' command", err), false)
+	}
+
+	resp, err := bloom.exists(args[1])
+	if err != nil {
+		return Encode(fmt.Errorf("%w for 'BFEXISTS' command", err), false)
+	}
+
+	return resp
+}
+
+// evalBFInfo evaluates the BFINFO command responsible for returning the
+// parameters and metadata of an existing bloom filter.
+func evalBFInfo(args []string) []byte {
+	if len(args) != 1 {
+		return Encode(fmt.Errorf("%w for 'BFINFO' command", errWrongArgs), false)
+	}
+
+	bloom, err := getOrCreateBloomFilter(args[0], nil)
+	if err != nil {
+		return Encode(fmt.Errorf("%w for 'BFINFO' command", err), false)
+	}
+
+	return Encode(bloom.info(args[0]), false)
+}
+
+// getOrCreateBloomFilter attempts to fetch an existing bloom filter from
+// the kv store. If it does not exist, it tries to create one with
+// given `opts` and returns it.
+func getOrCreateBloomFilter(key string, opts *BloomOpts) (*Bloom, error) {
+	obj := Get(key)
+
+	// If we don't have a filter yet and `opts` are provided, create one.
+	if obj == nil && opts != nil {
+		obj = NewObj(newBloomFilter(opts), -1, OBJ_TYPE_BITSET, OBJ_ENCODING_BF)
+		Put(key, obj)
+	}
+
+	// If no `opts` are provided for filter creation, return err
+	if obj == nil && opts == nil {
+		return nil, errInvalidKey
+	}
+
+	if err := assertType(obj.TypeEncoding, OBJ_TYPE_BITSET); err != nil {
+		return nil, err
+	}
+
+	if err := assertEncoding(obj.TypeEncoding, OBJ_ENCODING_BF); err != nil {
+		return nil, err
+	}
+
+	return obj.Value.(*Bloom), nil
+}
+
+// setBit sets the bit at index `b` to "1" in `buf`.
+func setBit(buf []byte, b int) {
+	if b < 0 {
+		return
+	}
+
+	idx, offset := b/8, 7-b%8
+	if idx < 0 || idx >= len(buf) {
+		return
+	}
+
+	buf[idx] = buf[idx] | 1<<offset
+}
+
+// isBitSet checks if the bit at index `b` is set to "1" or not in `buf`.
+func isBitSet(buf []byte, b int) bool {
+	if b < 0 {
+		return false
+	}
+
+	idx, offset := b/8, 7-b%8
+	if idx >= len(buf) {
+		return false
+	}
+
+	if buf[idx]&(1<<offset) == 1<<offset {
+		return true
+	} else {
+		return false
+	}
+}
```

Comment:
There is a util directory where you can put byte level utils; also we have something similar for `varint` and `bytelist`. Check that implementation once and see if we need to converge.

Reply:
Hi, at the moment, I don't think there's any code which can be converged. Bytelist holds some extra metadata and we might not need any of it for bloom bytes. Although I can move some logic to bloom_utils to separate the logic.